### DISCUSSION TITLE
More information about the structure of macad3d and how to write scripting of macad3d

### DISCUSSION THREAD

**Discussion by RileyGe** [Macad3D/Macad3D#11] [General]
Macad3d is a great project, and I hope I can participate in the development of macad3d. But I found that there is very little information about macad3d development or scripting. I hope the author can write more about the structure of macad3d and how to write scripting of macad3d.
Thanks again!

**Comment by dipts** [comment]
Hi RileyGe,
Thank you for your interest. There is an introduction to scripting in the User Guide, including examples for various use cases. You can also find them online [here](https://macad3d.net/userguide/html/dbef1051-1ced-4b54-89a5-64c02de1722d.htm). In addition, there is no developer documentation, but if you have a question or a specific request, I will be happy to answer it.

**Reply by RileyGe** [reply]
Hi Dipts,
I have read and tested all the examples in the document, they are very helpful to my understanding of Macad3D. But I want to directly modify the menu of Macad3D to generate new functions (including new Geometry), so I need to understand the code more deeply. Recently, I am also studying the code of Macad3D in depth. I may write some tutorials later, hoping to be helpful to those who study this project later.

**Reply by dipts** [reply]
Hi RileyGe,
scripts cannot yet have menu entries. In order for this to be useful, a possibility is required to execute scripts automatically when the program starts. Starting up Roslyn takes some time, so the compiled autostart scripts should be cached. The effort is manageable, but requires intensive testing. This approach can have advantages over implementation directly in Macad code, it depends a lot on the application.
